cli/module_build.go

```diff
@@ -44,3 +104,136 @@ func ModuleBuildLocalAction(c *cli.Context) error {
 	infof(c.App.Writer, "Completed build")
 	return nil
 }
+
+// ModuleBuildListAction lists the module's build jobs.
+func ModuleBuildListAction(c *cli.Context) error {
+	manifestPath := c.String(moduleFlagPath)
+	manifest, err := loadManifest(manifestPath)
+	if err != nil {
+		return err
+	}
+	var numberOfJobsToReturn *int32
+	if c.IsSet(moduleBuildFlagNumber) {
+		number := int32(c.Int(moduleBuildFlagNumber))
+		numberOfJobsToReturn = &number
+	}
+	var buildIDFilter *string
+	if c.IsSet(moduleBuildFlagBuildID) {
+		filter := c.String(moduleBuildFlagBuildID)
+		buildIDFilter = &filter
+	}
+	client, err := newViamClient(c)
+	if err != nil {
+		return err
+	}
+	moduleID, err := parseModuleID(manifest.ModuleID)
+	if err != nil {
+		return err
+	}
+	jobs, err := client.listModuleBuildJobs(moduleID, numberOfJobsToReturn, buildIDFilter)
+	if err != nil {
+		return err
+	}
+	// table format rules:
+	// minwidth, tabwidth, padding int, padchar byte, flags uint
+	w := tabwriter.NewWriter(c.App.Writer, 5, 4, 1, ' ', 0)
+	tableFormat := "%s\t%s\t%s\t%s\t%s\n"
+	fmt.Fprintf(w, tableFormat, "ID", "PLATFORM", "STATUS", "VERSION", "TIME")
+	for _, job := range jobs.Jobs {
+		fmt.Fprintf(w,
+			tableFormat,
+			job.BuildId,
+			job.Platform,
+			jobStatusFromProto(job.Status),
+			job.Version,
+			job.StartTime.AsTime().Format(time.RFC3339))
+	}
+	//nolint: errcheck,gosec
+	w.Flush()
+	return nil
+}
+
+// ModuleBuildLogsAction retrieves the logs for a specific build step.
+func ModuleBuildLogsAction(c *cli.Context) error {
+	buildID := c.String(moduleBuildFlagBuildID)
+	platform := c.String(moduleBuildFlagPlatform)
+	shouldWait := c.Bool(moduleBuildFlagWait)
+	if shouldWait {
+		return errors.New("wait not implemented")
+	}
+
+	client, err := newViamClient(c)
+	if err != nil {
+		return err
+	}
+
+	err = client.printModuleBuildLogs(c.Context, buildID, platform)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *viamClient) printModuleBuildLogs(ctx context.Context, buildID, platform string) error {
+	if err := c.ensureLoggedIn(); err != nil {
+		return err
+	}
+
+	logsReq := &buildpb.GetLogsRequest{
+		BuildId:  buildID,
+		Platform: platform,
+	}
+
+	stream, err := c.buildClient.GetLogs(c.c.Context, logsReq)
+	if err != nil {
+		return err
+	}
+	lastBuildStep := ""
+	for {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		log, err := stream.Recv()
```

Comment:
I realized that it is quite surprising that it doesn't take a context. After looking it up, it seems like this is a known issue and the fix is to just put it in a goroutine :facepalm: https://github.com/grpc/grpc-go/issues/1229#issuecomment-300938770

Reply:
>stream.RecvCtx(...) or some stream.SetTimeout(...) method to control the deadline of the Recv call.

That is what I was hoping to find, but I can't seem to find it.

Reply:
I've put a basic impl in the bottom of f1d120d68

It is a bit inefficient because it creates a new goroutine every loop, but from what I've been reading about goroutines, that seems cheap compared to network io.

If you have a better impl, I would love to see what you come up with.

Reply:
Reverted to `Recv` because I /think/ it respects the ctx of the stream which is the ctx of the cli (passed to GetLogs).

I haven't tested this and it is many layers deep in grpc code, but I am ~60% confident. (also the rest of viam uses the same pattern so if it is a problem, it is a problem everywhere. so we should fix it then)

```grpc/internal/transport/transport.go:321```
```
func (s *Stream) waitOnHeader() {
	if s.headerChan == nil {
		// On the server headerChan is always nil since a stream originates
		// only after having received headers.
		return
	}
	select {
	case <-s.ctx.Done():
		// Close the stream to prevent headers/trailers from changing after
		// this function returns.
		s.ct.CloseStream(s, ContextErr(s.ctx.Err()))
		// headerChan could possibly not be closed yet if closeStream raced
		// with operateHeaders; wait until it is closed explicitly here.
		<-s.headerChan
	case <-s.headerChan:
	}
}

// RecvCompress returns the compression algorithm applied to the inbound
// message. It is empty string if there is no compression applied.
func (s *Stream) RecvCompress() string {
	s.waitOnHeader()
	return s.recvCompress
}
```

Reply:
Cool; after offline discussion, `stream.Recv()` seems totally fine for now (especially given, as you pointed out, the prevalence of no-timeout-network-IO across the CLI). Thanks for digging into this; I'll trust the 60% for now and we can reconsider if NetCode and/or fleet decide we need sensible default timeouts for the CLI.